pkcs11/src/new/types/object.rs

```diff
@@ -171,27 +235,133 @@ impl Attribute<'_> {
             | Attribute::Verify(b)
             | Attribute::VerifyRecover(b)
-            | Attribute::Wrap(b) => *b as *mut _ as CK_VOID_PTR,
-            Attribute::Base(byte) => *byte as *mut _ as CK_VOID_PTR,
-            Attribute::Class(object_class) => *object_class as CK_OBJECT_CLASS_PTR as CK_VOID_PTR,
-            Attribute::KeyType(key_type) => *key_type as *mut _ as CK_VOID_PTR,
-            Attribute::Label(label) => label.as_ptr() as CK_VOID_PTR,
-            Attribute::ModulusBits(bits) => *bits as *mut _ as CK_VOID_PTR,
-            Attribute::Prime(bytes) => bytes.as_ptr() as CK_VOID_PTR,
-            Attribute::PublicExponent(bytes) => bytes.as_ptr() as CK_VOID_PTR,
-            Attribute::Modulus(bytes) => bytes.as_ptr() as CK_VOID_PTR,
-            Attribute::Value(value) => value.as_ptr() as CK_VOID_PTR,
-            Attribute::ValueLen(len) => *len as *mut _ as CK_VOID_PTR,
+            | Attribute::Wrap(b) => b.as_ref().get_ref() as *const _ as *mut c_void,
+            // CK_ULONG
+            Attribute::ModulusBits(val) | Attribute::ValueLen(val) => {
+                val.as_ref().get_ref() as *const _ as *mut c_void
+            }
+            // Vec<u8>
+            Attribute::Base(bytes)
+            | Attribute::Label(bytes)
+            | Attribute::Prime(bytes)
+            | Attribute::PublicExponent(bytes)
+            | Attribute::Modulus(bytes)
+            | Attribute::Value(bytes)
+            | Attribute::Id(bytes) => bytes.as_ptr() as *mut c_void,
+            // Unique types
+            Attribute::Class(object_class) => {
+                object_class.as_ref().get_ref() as *const _ as *mut c_void
+            }
+            Attribute::KeyType(key_type) => key_type.as_ref().get_ref() as *const _ as *mut c_void,
+            Attribute::AllowedMechanisms(mechanisms) => {
+                mechanisms.as_ref().get_ref() as *const _ as *mut c_void
+            }
         }
     }
 }
 
-impl From<&mut Attribute<'_>> for CK_ATTRIBUTE {
-    fn from(attribute: &mut Attribute) -> Self {
+impl From<&Attribute> for CK_ATTRIBUTE {
+    fn from(attribute: &Attribute) -> Self {
         Self {
             type_: attribute.attribute_type().into(),
             pValue: attribute.ptr(),
-            ulValueLen: attribute.len(),
+            // Truncation from usize to CK_ULONG not checked
+            // Should be fine in most cases.
+            ulValueLen: attribute.len() as CK_ULONG,
+        }
+    }
+}
+
+impl TryFrom<CK_ATTRIBUTE> for Attribute {
+    type Error = Error;
+
+    fn try_from(attribute: CK_ATTRIBUTE) -> Result<Self, Error> {
+        let attr_type = AttributeType::try_from(attribute.type_)?;
+        // Cast from c_void to u8
+        let val = unsafe {
+            std::slice::from_raw_parts(
+                attribute.pValue as *const u8,
+                attribute.ulValueLen.try_into()?,
+            )
+        };
+        match attr_type {
+            // CK_BBOOL
```

Comment:
I am not really happy with this conversion. It will really depend on how we decide to create new `Attribute`: either with one constructor per value, one constructor/type per `CK_` type, etc...

Reply:
You could have either a macro or some private functions defined in this file to do the common conversions? A macro makes the code a bit harder to follow when you first look at it, but maybe easier to not make mistakes

Reply:
Yes I guess a macro would be nice here. Still I'd like to use in it the constructor of the abstracted types (`CK_BBOOL`, etc) instead of what is currently there.